vsintegration/src/FSharp.Editor/AutomaticCompletion/BraceCompletionSessionProvider.fs

```diff
@@ -0,0 +1,561 @@
+        match openingBrace with
+        | Parenthesis.OpenCharacter -> ParenthesisCompletionSession() :> IEditorBraceCompletionSession
+        | CurlyBrackets.OpenCharacter -> ParenthesisCompletionSession() :> IEditorBraceCompletionSession
+        | SquareBrackets.OpenCharacter -> ParenthesisCompletionSession() :> IEditorBraceCompletionSession
+        | VerticalBar.OpenCharacter -> VerticalBarCompletionSession() :> IEditorBraceCompletionSession
+        | AngleBrackets.OpenCharacter -> AngleBracketCompletionSession() :> IEditorBraceCompletionSession
+        | DoubleQuote.OpenCharacter -> DoubleQuoteCompletionSession() :> IEditorBraceCompletionSession
+        | Asterisk.OpenCharacter -> DoubleQuoteCompletionSession() :> IEditorBraceCompletionSession
+        | _ -> null
+
+    interface IEditorBraceCompletionSessionFactory with
+
+        member this.TryCreateSession(document, openingPosition, openingBrace, cancellationToken) = 
+            if this.IsSupportedOpeningBrace(openingBrace) && this.CheckCodeContext(document, openingPosition, openingBrace, cancellationToken) then
+                this.CreateEditorSession(document, openingPosition, openingBrace, cancellationToken)
+            else
+                null
+
+[<Export(typeof<IBraceCompletionSessionProvider>)>]
+[<ContentType(FSharpConstants.FSharpContentTypeName)>]
+[<BracePair(Parenthesis.OpenCharacter, Parenthesis.CloseCharacter)>]
+[<BracePair(CurlyBrackets.OpenCharacter, CurlyBrackets.CloseCharacter)>]
+[<BracePair(SquareBrackets.OpenCharacter, SquareBrackets.CloseCharacter)>]
+[<BracePair(DoubleQuote.OpenCharacter, DoubleQuote.CloseCharacter)>]
+[<BracePair(VerticalBar.OpenCharacter, VerticalBar.CloseCharacter)>]
+[<BracePair(AngleBrackets.OpenCharacter, AngleBrackets.CloseCharacter)>]
+[<BracePair(Asterisk.OpenCharacter, Asterisk.CloseCharacter)>]
+type BraceCompletionSessionProvider
+    [<ImportingConstructor>] 
+    (
+        undoManager: ITextBufferUndoManagerProvider,
+        editorOperationsFactoryService: IEditorOperationsFactoryService
+    ) =
+
+    interface IBraceCompletionSessionProvider with
+
+        member __.TryCreateSession(textView, openingPoint, openingBrace, closingBrace, session) =
+            let textSnapshot = openingPoint.Snapshot
+
+            let newSession =
+                match textSnapshot.GetOpenDocumentInCurrentContextWithChanges() with
+                | null -> null
+                | document ->
+                    match getLanguageService<IEditorBraceCompletionSessionFactory> document with
+                    | null -> null
+                    | editorSessionFactory ->
+                        // Brace completion is (currently) not cancellable.
+                        let cancellationToken = CancellationToken.None
+                        
+                        match editorSessionFactory.TryCreateSession(document, openingPoint.Position, openingBrace, cancellationToken) with
+                        | null -> null
```

Comment:
Maybe `maybe { }` instead of all these nested matches?

Reply:
Let's see what we can do.

Reply:
I'm not saying that using maybe CE would be a clear win. It'd remove the nested matches, but it'd make debugging experience _a lot_ worse. So, maybe leaving it as is is the best we can do :(

Reply:
Normally I stay away from CEs, but I'm open to use it here because of the amount of nesting. If I were to use a CE here, I would need to add an overload to `maybe`:

```
    [<DebuggerStepThrough>]
    member inline __.Bind(value: 'T, f: 'T -> 'U option) : 'U option =
        if value = null then
            None
        else
            Some value
        |> Option.bind f
```

This way I could handle types that can be null and are not option types.

Reply:
But, I'm not sure this is a good idea to do. I figure this increases complexity for `maybe`.

Reply:
Can't you just pipe into `Option.ofObj` instead?

Reply:
`Option.ofObj` is better than adding an overload, didn't know we had it.

Ok, so we are using `maybe`. It does look cleaner than the nesting, I'm ok with this. Thank you guys.